src/Fmt_ast.ml

```diff
@@ -34,6 +34,54 @@ let protect =
         first := false ) ;
       raise exc
 
+let is_disabled = function
+  | {txt= "ocamlformat.disable"}, _ -> true
+  | _ -> false
+
+let scope c a =
+  let error {loc; txt} reason =
+    let w = Warnings.Attribute_payload (txt, reason) in
+    !Location.warning_printer loc Caml.Format.err_formatter w ;
+    c
+  in
+  let misplaced {loc; txt} =
+    let w = Warnings.Misplaced_attribute txt in
+    !Location.warning_printer loc Caml.Format.err_formatter w ;
+    c
+  in
+  match a with
+  | ({txt= "ocamlformat"} as aname), payload -> (
+    match payload with
+    | PStr
+        [ { pstr_desc=
+              Pstr_eval
+                ( { pexp_desc= Pexp_constant (Pconst_string (str, None))
+                  ; pexp_attributes= [] }
+                , [] ) } ] ->
+        List.fold ~init:c (String.split ~on:';' str) ~f:(fun c s ->
+            match String.lsplit2 ~on:'=' s with
+            | Some (name, value) -> (
+              try
+                Conf.update c ~name:(String.strip name)
+                  ~value:(String.strip value)
+              with e -> error aname (Exn.to_string e) )
+            | None ->
+                error aname "malformat option, should be [name=value]+" )
+    | _ -> error aname "string expected" )
+  | ({txt= "ocamlformat.disable"} as aname), _ -> misplaced aname
+  | ({txt} as aname), _ when String.is_prefix ~prefix:"ocamlformat." txt ->
+      error aname
+        (Format.sprintf "unknown suffix %S"
+           (String.chop_prefix_exn ~prefix:"ocamlformat." txt))
+  | _ -> c
+
+let scope c l = {c with conf= List.fold ~init:c.conf l ~f:scope}
+
+let maybe_disabled c attrs loc f =
+  if List.exists attrs ~f:is_disabled then
+    str (Source.string_at c.source loc)
```

Comment:
This will need to remove the comments attached to the locations in the skipped part of the parsetree, right?

Reply:
yes

Reply:
now done